Protected-FileSystem(PFS) library

sudharkrish opened the pull request:
### **This library is a plugin for Graphene, that provides file-system shield to Graphene, by incorporating file protection feature from Intel's SGX-SDK's Protected-FS library.**

### Brief overview of PR, with info on directories and components added:

1. **docs folder** has design document. Note: Code is NOT fully upto the design document, CHANGELOG has pending features.
With GPFS library,  the features that have been implemented are: filename confidentiality, filename integrity, file-path integrity and binding of files to a volume,
on top of existing filesystem protection software (i.e. SGX-Protected Filesystem).

2. **ld_preload_lib**: Has sources to intercept and overload file-system apis, and route it to SGX ProtectedFS library or to C filesystem api. Also has sources to augment cryptographic mechanisms
to add additional features as described in the design doc under docs directory.

3. **pfs_sdk**: Has a patch file that gets applied on SGX SDK's protected-fs library, to make it work within graphene, without sgx-sdk dependencies.

4. **pfs_proxy**: Has changes, to hook calls from SGX SDK-protectefs library sources, to non-sgx-sdk apis.

5. **apps/protfs_app**: Has sources for sample graphene application, that sets the following in the
application's manifest(specifies path to proteced directory, sets the overloaded GPFS library, sets the type of key to be used for file-protection), and uses GPFS library
transparently as a file-system shield

### For more details, refer to README.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/oscarlab/graphene-plugins/1)
<!-- Reviewable:end -->

skris14 committed:
Graphene-ProtectedFileSystem shield(GPFS) library:
This library is a plugin for Graphene, that provides file-system shield to Graphene, by incorporating file protection feature from
Intel's SGX-SDK's Protected-FS library.
For more details, refer to README

mkow reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene-plugins/1)* status: 0 of 53 files reviewed, 2 unresolved discussions (waiting on @sudharkrish)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene-plugins/1#-LmeM5XlAnuQzv5KwkGh:-LmeM5XlAnuQzv5KwkGi:b-mrq5j1) (no related file):*
Please fix the email address in your commit so that GitHub can recognize it (either by changing it in the commit or by adding it to your GitHub account). You can push-force if you choose the former.

---
*[a discussion](https://reviewable.io/reviews/oscarlab/graphene-plugins/1#-LmeMGmU0BJslkxDhtkH:-LmeMGmU0BJslkxDhtkI:b-p3umu8) (no related file):*
Why the directory is called "protect-fs" when the plugin is named "protected-fs"? The naming in other places is also inconsistent, like "protfs" and "pfs".

---


<!-- Sent from Reviewable.io -->

mkow reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene-plugins/1)* status: 0 of 53 files reviewed, 2 unresolved discussions (waiting on @sudharkrish)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene-plugins/1#-LmeMGmU0BJslkxDhtkH:-LmeOkfT4Z5WUEXvtlQd:bx3d42s) (no related file):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

Why the directory is called "protect-fs" when the plugin is named "protected-fs"? The naming in other places is also inconsistent, like "protfs" and "pfs".
</blockquote></details>

There's also "Graphene-ProtectedFileSystem shield", "GPFS" and "Protected-FS". Please unify the naming.

---


<!-- Sent from Reviewable.io -->

sudharkrish reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene-plugins/1)* status: 0 of 53 files reviewed, 2 unresolved discussions (waiting on @mkow)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene-plugins/1#-LmeM5XlAnuQzv5KwkGh:-Lmfc9ne3U15y348JU52:b-5hhnwo) (no related file):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

Please fix the email address in your commit so that GitHub can recognize it (either by changing it in the commit or by adding it to your GitHub account). You can push-force if you choose the former.
</blockquote></details>

Completed. I have added my email to GitHub account. 

---
*[a discussion](https://reviewable.io/reviews/oscarlab/graphene-plugins/1#-LmeMGmU0BJslkxDhtkH:-LmfcyySCVewHpFROSEv:b-75pel1) (no related file):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

There's also "Graphene-ProtectedFileSystem shield", "GPFS" and "Protected-FS". Please unify the naming.
</blockquote></details>

I agree that having the same convention across, will be better. For example, I can stick to acronym->pfs, which stands for protected-filesystem. The only concern is, if I stick to pfs everywhere, the only place to find out, what pfs stands for, would be the README file. So in some places, like the directory names, it may be better to have partial expansion, like protected-fs. I am using the acronym pfs in code. Option1->Keep directory names as protected-fs, but we can keep the acronym->pfs in other places like code. Option2->  Keep pfs everywhere(directory-names, code, README file), in this case README will have expansion for what PFS stands for.   You can let me know, if you have a preference for Option1 or Option2.

---


<!-- Sent from Reviewable.io -->

mkow reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene-plugins/1)* status: 0 of 53 files reviewed, 1 unresolved discussion (waiting on @sudharkrish)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene-plugins/1#-LmeM5XlAnuQzv5KwkGh:-LmfolK91nMsHtcwOskM:b-uz91tr) (no related file):*
<details><summary><i>Previously, sudharkrish wrote…</i></summary><blockquote>

Completed. I have added my email to GitHub account. 
</blockquote></details>

Looks good now, thanks

---
*[a discussion](https://reviewable.io/reviews/oscarlab/graphene-plugins/1#-LmeMGmU0BJslkxDhtkH:-LmfooruEmifbBueTCUI:b-k7lipn) (no related file):*
<details><summary><i>Previously, sudharkrish wrote…</i></summary><blockquote>

I agree that having the same convention across, will be better. For example, I can stick to acronym->pfs, which stands for protected-filesystem. The only concern is, if I stick to pfs everywhere, the only place to find out, what pfs stands for, would be the README file. So in some places, like the directory names, it may be better to have partial expansion, like protected-fs. I am using the acronym pfs in code. Option1->Keep directory names as protected-fs, but we can keep the acronym->pfs in other places like code. Option2->  Keep pfs everywhere(directory-names, code, README file), in this case README will have expansion for what PFS stands for.   You can let me know, if you have a preference for Option1 or Option2.
</blockquote></details>

Option 1 sounds better to me, but I'd prefer to restrict using `pfs` to variable and function names only, e.g. when prefixing them like `pfs_read_file`, so for example comments in the code would use "protected-fs" name (or "Protected-FS" - please decide about the casing and dashing also and make this consistent).

---


<!-- Sent from Reviewable.io -->

sudharkrish reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene-plugins/1)* status: 0 of 53 files reviewed, 1 unresolved discussion (waiting on @mkow)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene-plugins/1#-LmeMGmU0BJslkxDhtkH:-LmgEpJoDKP5Z6jei4Bk:b-l6oe39) (no related file):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

Option 1 sounds better to me, but I'd prefer to restrict using `pfs` to variable and function names only, e.g. when prefixing them like `pfs_read_file`, so for example comments in the code would use "protected-fs" name (or "Protected-FS" - please decide about the casing and dashing also and make this consistent).
</blockquote></details>

Sure, sounds good. Will do the change, per Option1, along with your suggestions.

---


<!-- Sent from Reviewable.io -->

skris14 committed:
Based on initial review comment->
1. Changes to make PFS acronym(which stands for Protected-Filesystem) consistent
across directory-names, code, README and application sources.
2. Also fixed few issues in pfs_app test sources.

sudharkrish committed:
Merge 261cc614624b92a704993c7ba3a452c9316f8246 into 0f277912ac899f719491a09f487db24502c7603e

sudharkrish commented:
@mkow updated PR, to make acronym->PFS consistent across directory-names, code, README and application sources(pfs_app). 
Link to updated commit of PR: https://github.com/oscarlab/graphene-plugins/pull/1/commits/261cc614624b92a704993c7ba3a452c9316f8246

skris14 committed:
Changes to make PFS SDK library C-based, and removed C++ Dependencies. In order to build this, will need following 2 dependencies: - Hashmap Lib: Using the hashmap library(MIT License, same as mbedtls
 used by Graphene): You can pull the sources from here-> https://github.com/DavidLeeds/hashmap - Also needs list.h in Graphene-PAL, to have few additional features like sorting, and extra field like size.
 Please refer to changes in just 1 file->list.h. https://github.com/oscarlab/graphene/pull/1268
